fix: handle NaN and Infinity float values in gRPC

ariasmn opened the pull request:
TBD

ariasmn commented:
Couple of things:

- Having a hard time trying to understand how to build the tests. I just checked that current tests pass, but I'd need a bit of help to build a test case for this, sorry about that.
- I still don't think that this approach is good enough... I feel like maybe the best bet would be to create the proto message using `protoreflect` and then marshaling normally, but I haven't tried if that would break wrappers and reflection implementations.

mstoykov reviewed (commented):
Thank you for the PR and sorry for the slow reply :bow: 

I have left some comments around the code. 

On the whole idea I am still not particularly certain this is ... what should be done, if anything. 

Trying to figure out how this works in other cases I found https://github.com/protocolbuffers/protobuf-javascript/issues/49 which seems to suggest that this isn't really supported. But maybe I am getting something wrong.

From the test it seems that this just works on the way back which seems awfully convient.

ariasmn commented on internal/js/modules/k6/grpc/client.go:
```diff
@@ -360,7 +361,14 @@ func (c *Client) buildInvokeRequest(
-	b, err := req.ToObject(c.vu.Runtime()).MarshalJSON()
+
+	object := req.ToObject(c.vu.Runtime())
+	normalized, err := normalizeNumberStrings(object, c.vu.Runtime())
```
But the object is built based on the `req`, which is a `sobek.Value` passed by value, so we shouldn't worry about it, no? Or am I misunderstanding you? (Probably this, my brain is fried today, sorry :disappointed: )

ariasmn replied:
Forget about my last comment, I took some time today (it's been a while, sorry :disappointed: ) and completely understood what you meant.

Don't know if the approach I did is good enough or if there's a better way... We can discuss it in another conversation.

ariasmn commented:
Hey @mstoykov, thanks a lot for the review!

I’ve made some changes based on your comments. There’s one part I didn't change yet because I’m not fully sure I understand it correctly, but we can keep discussing it in the open conversation.

Regarding the issue you mentioned: it looks like an older open issue. From what I saw, the code has since changed and now properly asserts both `Infinity` and `NaN`:

https://github.com/protocolbuffers/protobuf-javascript/blob/0768cc9f18bec192c393d7c2962ee8ab5ad1931b/binary/encoder.js#L411-L418

I also tested it locally using the dummy.proto from the issue, for both client and server (wire format). Additionally, the [ProtoJSON documentation](https://protobuf.dev/programming-guides/json/) confirms that both values are officially supported.

Just a heads-up: I might be a bit slower next week if any further changes are needed — apologies in advance.

Thanks again for the great feedback. 🙏

ariasmn commented:
Hi there,

sorry for the delay on this! It's been a while since I could get back to this PR (I haven't had much time recently).

I've made the requested corrections. Please take a look and let me know if everything looks good (I feel like there should be a better way of handling the copy for sure) or if there's anything else that needs to be addressed.

Thank you! 🙇

ariasmn commented:
Hi @joanlopez,

Thank you for catching that one! I implemented a fix by following what's done in the `stringify` method in Sobek, and added a test case as well:

https://github.com/grafana/k6/blob/602467dee735ba74fc9664b23ca53220f1212f0c/vendor/github.com/grafana/sobek/builtin_json.go#L366-L372

Let me know if my choice of wording for variables/comments is odd, I can fix them no problem.

Thanks!

joanlopez commented on internal/js/modules/k6/grpc/client.go:
```diff
@@ -387,6 +397,71 @@ func (c *Client) buildInvokeRequest(
+	}
+
+	// Handle special float values.
+	if f, ok := v.Export().(float64); ok {
```
As @mstoykov suggested before, could we use `ExportType()` here instead of `Export()`, as the latter is a much more expensive operation.

In case it's a `float64`, I guess you'll need to call `v.Export()` anyway, but in all the other cases, it won't be called that early.

joanlopez replied:
Or maybe even something like:

```suggestion
	if v.ExportType().Kind() == reflect.Float64 {
		f := v.ToFloat()
		switch {
		case math.IsNaN(f):
			return runtime.ToValue("NaN"), nil
		case math.IsInf(f, 1):
			return runtime.ToValue("Infinity"), nil
		case math.IsInf(f, -1):
			return runtime.ToValue("-Infinity"), nil
		}

		return v, nil
	}
```

So, you don't need `Export()` at all.
Wdyt? @mstoykov

mstoykov replied:
This seems better.

For the record - the idea is that if you are using `Export` it will do a lot of work in some of the cases. If the end type is flaot64 - that probably is fine. But it also might be a big object that you just fully transformed from JS to go type. And then you just throw it away because it isn't a float64, just to repeat the operation each time.

joanlopez replied:
Sure, no problem @ariasmn, that's why we have a review process, we're all humans and sometimes it happens!

Looks good now! 👌🏻 Thanks! 🙌🏻

mstoykov reviewed (approved):
LGTM! Thanks for the changes and going through figuring it out @ariasmn :bow:
